src/Anoma/Effect/Base.hs

```diff
@@ -0,0 +1,150 @@
+-- | This module assumes that the following external dependencies are installed:
+-- 1. mix
+--
+-- 2. grpcurl
+module Anoma.Effect.Base
+  ( Anoma,
+    noHalt,
+    anomaRpc,
+    AnomaPath (..),
+    anomaPath,
+    runAnoma,
+    module Anoma.Rpc.Base,
+    module Juvix.Compiler.Nockma.Translation.FromTree,
+  )
+where
+
+import Anoma.Effect.Paths
+import Anoma.Rpc.Base
+import Data.ByteString qualified as B
+import Juvix.Compiler.Nockma.Translation.FromTree (AnomaResult)
+import Juvix.Data.CodeAnn
+import Juvix.Extra.Paths (anomaStartExs)
+import Juvix.Prelude
+import Juvix.Prelude.Aeson (Value, eitherDecodeStrict, encode)
+
+data Anoma :: Effect where
+  -- | Keep the node and client running
+  NoHalt :: Anoma m ExitCode
+  -- | grpc call
+  AnomaRpc :: GrpcMethodUrl -> Value -> Anoma m Value
+
+makeSem ''Anoma
+
+newtype AnomaPath = AnomaPath {_anomaPath :: Path Abs Dir}
+
+newtype GrpcPort = GrpcPort {_grpcPort :: Int}
+
+makeLenses ''AnomaPath
+makeLenses ''GrpcPort
+
+listenPort :: Int
+listenPort = 50051
+
+relativeToAnomaDir :: (Members '[Reader AnomaPath] r) => Path Rel x -> Sem r (Path Abs x)
+relativeToAnomaDir p = do
+  anoma <- asks (^. anomaPath)
+  return (anoma <//> p)
+
+withSpawnAnomaClient ::
+  (Members '[Process, Logger, EmbedIO, Reader AnomaPath, Reader GrpcPort, Error SimpleError] r) =>
+  (ProcessHandle -> Sem r a) ->
+  Sem r a
+withSpawnAnomaClient body = do
+  cprocess <- mkProcess
+  withCreateProcess cprocess $ \_stdin mstdout _stderr procHandle -> do
+    let out = fromJust mstdout
+    txt <- hGetLine out
+    case takeWhile (/= '.') (unpack txt) of
+      "Connected to node" -> do
+        logInfo "Anoma client successfully started"
+        logInfo (mkAnsiText ("Listening on port " <> annotate AnnImportant (pretty listenPort)))
+        body procHandle
+      _ -> throw (SimpleError (mkAnsiText @Text "Something went wrong when starting the anoma client"))
+  where
+    mkProcess :: (Members '[Reader AnomaPath, Reader GrpcPort] r') => Sem r' CreateProcess
+    mkProcess = do
+      grpcport <- asks (^. grpcPort)
+      anomaClient <- relativeToAnomaDir clientRelFile
+      return
+        ( proc
+            (toFilePath anomaClient)
+            [ "--listen-port",
+              show listenPort,
+              "--node-host",
+              "localhost",
+              "--node-port",
+              show grpcport
+            ]
+        )
+          { std_out = CreatePipe
+          }
+
+withSapwnAnomaNode ::
+  forall r a.
+  (Members '[EmbedIO, Logger, Error SimpleError, Process, Reader AnomaPath] r) =>
+  (Int -> Handle -> ProcessHandle -> Sem r a) ->
+  Sem r a
+withSapwnAnomaNode body = withSystemTempFile "start.exs" $ \fp tmpHandle -> do
+  liftIO (B.hPutStr tmpHandle anomaStartExs)
+  hClose tmpHandle
+  curDir <- getCurrentDir
+  asks (^. anomaPath) >>= setCurrentDir
```

Comment:
suggestion: Use `anomaPath` directly to make an absolute path for the process working directory. 

Use of `setCurrentDir` will cause issues if this function is ever used concurrently in the future so we should try to avoid using it in general.

Reply:
fixed in c8872e0542b109621f0dfbc19ccb4cd96fad7f9f

Reply:
The calls to `setCurrentDir` are still there. Can they be removed now the cwd is being set on CreateProcess?

Reply:
I forgot to remove those :disappointed: it is fixed now